Mathlib/Tactic/Linter/DocModule.lean

```diff
@@ -0,0 +1,210 @@
+    let args := args
+    let first := args.get? 0
+    first.map (·.isOfKind ``Lean.Parser.Command.moduleDoc)
+  | _=> some false
+
+/-- returns the array of `import` identifiers. -/
+partial
+def getImportIds (s : Syntax) : Array Syntax :=
+  let rest : Array Syntax := (s.getArgs.map getImportIds).flatten
+  if s.isOfKind ``Lean.Parser.Module.import then
+    rest.push (s.getArgs.getD 2 default)
+  else
+    rest
+
+/-- `parseUpToHere stx post` takes as input a `Syntax` `stx` and a `String` `post`.
+It parses the file containing `stx` up to and excluding `stx`, appending `post` at the end.
+
+The option of appending a final string to the text gives more control to avoid syntax errors,
+for instance in the presence of `#guard_msgs in` or `set_option ... in`.
+-/
+def parseUpToHere (stx : Syntax) (post : String := "") (included : Bool := true) :
+    CommandElabM Syntax := do
+  let fm ← getFileMap
+  let startPos := if included then stx.getTailPos?.getD default else stx.getPos?.getD default
+  let upToHere : Substring:= { str := fm.source, startPos := ⟨0⟩, stopPos := startPos}
+  -- append a further string after the `upToHere` content
+  Parser.testParseModule (← getEnv) "linter.style.header" (upToHere.toString ++ post)
+
+/-- `toSyntax s pattern` converts the two input strings into a `Syntax`, assuming that `pattern`
+is a substring of `s`:
+the syntax is an atom with value `pattern` whose the range is the range of `pattern` in `s`. -/
+def toSyntax (s pattern : String) (offset : String.Pos := 0) : Syntax :=
+  --let firstSubstring := ((s.splitOn pattern).getD (1 + offset) "").toSubstring
+  let beg := ((s.splitOn pattern).getD 0 "").endPos + offset
+  let fin := (((s.splitOn pattern).getD 0 "") ++ pattern).endPos + offset
+  mkAtomFrom (.ofRange ⟨beg, fin⟩) pattern
+
+-- adapted from the textbased
+/-- Return if `line` looks like a correct authors line in a copyright header. -/
+def authorsLineCorrections (line : String) (offset : String.Pos) : Array (Syntax × String) :=
+  Id.run do
+  -- We cannot reasonably validate the author names, so we look only for a few common mistakes:
+  -- the file starting wrong, double spaces, using ' and ' between names,
+  -- and ending the line with a period.
+  let mut stxs := #[]
+  if !line.startsWith "Authors: " then
+    stxs := stxs.push
+      (toSyntax line (line.take "Authors: ".length) offset,
+       s!"The authors line should begin with 'Authors: '")
+  if ((line.replace "\n  " " ").splitOn "  ").length != 1 then
+    stxs := stxs.push (toSyntax line "  " offset, s!"Double spaces are not allowed.")
+  if (line.splitOn " and ").length != 1 then
+    stxs := stxs.push (toSyntax line " and " offset, s!"Please, do not use 'and', use ',' instead.")
+  if line.endsWith "." then
+    stxs := stxs.push
+      (toSyntax line "." offset,
+       s!"Please, do not end the authors' line with a period.")
+  return stxs
+
+/-- The main function to validate the copyright string. -/
```

Comment:
Can you document the return value as well?

Reply:
Documentation added!